Adding How To Guides for .NET for Apache Spark

Niharikadutta opened the pull request:
## Summary

This PR adds documents (How-to Guides) for [.NET for Apache Spark](https://github.com/dotnet/spark) for the following topics:

- [x] Connecting to Remote Storage (ADLS and WASB) from your Local Machine
- [x] Connecting to Data Sources
  - [x] SQL Server
  - [x] Event Hub
  - [x] MongoDB
- [x] User-defined Functions in .NET for Apache Spark (Invoking Scala/Java UDFs from Spark .NET apps and vice versa)
- [x] Running UDFs in notebooks

This PR addresses part of the following [.NET for Apache Spark issue](https://github.com/dotnet/spark/issues/671)

Preview Links:


File Type | File Name | Published Url
-- | -- | --
Content | docs/spark/how-to-guides/connect-to-azure-storage.md | https://review.docs.microsoft.com/en-us/dotnet/spark/how-to-guides/connect-to-azure-storage?branch=pr-en-us-20778
Content | docs/spark/how-to-guides/connect-to-event-hub.md | https://review.docs.microsoft.com/en-us/dotnet/spark/how-to-guides/connect-to-event-hub?branch=pr-en-us-20778
Content | docs/spark/how-to-guides/connect-to-mongo-db.md | https://review.docs.microsoft.com/en-us/dotnet/spark/how-to-guides/connect-to-mongo-db?branch=pr-en-us-20778
Content | docs/spark/how-to-guides/connect-to-sql-server.md | https://review.docs.microsoft.com/en-us/dotnet/spark/how-to-guides/connect-to-sql-server?branch=pr-en-us-20778
Content | docs/spark/how-to-guides/dotnet-interactive-udf-issue.md | https://review.docs.microsoft.com/en-us/dotnet/spark/how-to-guides/dotnet-interactive-udf-issue?branch=pr-en-us-20778
Content | docs/spark/how-to-guides/java-udf-from-dotnet.md | https://review.docs.microsoft.com/en-us/dotnet/spark/how-to-guides/java-udf-from-dotnet?branch=pr-en-us-20778
Resource | docs/spark/how-to-guides/media/connect-external-sources/hadoop-env.png | https://review.docs.microsoft.com/en-us/dotnet/spark/how-to-guides/media/connect-external-sources/hadoop-env.png?branch=pr-en-us-20778
Resource | docs/spark/how-to-guides/media/connect-external-sources/mongo-code.png | https://review.docs.microsoft.com/en-us/dotnet/spark/how-to-guides/media/connect-external-sources/mongo-code.png?branch=pr-en-us-20778
Resource | docs/spark/how-to-guides/media/connect-external-sources/SqlServerAuth.png | https://review.docs.microsoft.com/en-us/dotnet/spark/how-to-guides/media/connect-external-sources/sqlserverauth.png?branch=pr-en-us-20778
Resource | docs/spark/how-to-guides/media/connect-external-sources/SQLServerTCPIIPPort.png | https://review.docs.microsoft.com/en-us/dotnet/spark/how-to-guides/media/connect-external-sources/sqlservertcpiipport.png?branch=pr-en-us-20778
Resource | docs/spark/how-to-guides/media/connect-external-sources/SqlServerTCPIP.png | https://review.docs.microsoft.com/en-us/dotnet/spark/how-to-guides/media/connect-external-sources/sqlservertcpip.png?branch=pr-en-us-20778
Resource | docs/spark/how-to-guides/media/dotnet-interactive/broadcast-fails.png | https://review.docs.microsoft.com/en-us/dotnet/spark/how-to-guides/media/dotnet-interactive/broadcast-fails.png?branch=pr-en-us-20778
Resource | docs/spark/how-to-guides/media/dotnet-interactive/decompiledAssembly.png | https://review.docs.microsoft.com/en-us/dotnet/spark/how-to-guides/media/dotnet-interactive/decompiledassembly.png?branch=pr-en-us-20778
Resource | docs/spark/how-to-guides/media/dotnet-interactive/dotnet-interactive-cells.png | https://review.docs.microsoft.com/en-us/dotnet/spark/how-to-guides/media/dotnet-interactive/dotnet-interactive-cells.png?branch=pr-en-us-20778
Resource | docs/spark/how-to-guides/media/dotnet-interactive/udf_fails.png | https://review.docs.microsoft.com/en-us/dotnet/spark/how-to-guides/media/dotnet-interactive/udf_fails.png?branch=pr-en-us-20778
Resource | docs/spark/how-to-guides/media/dotnet-interactive/udf_works.png | https://review.docs.microsoft.com/en-us/dotnet/spark/how-to-guides/media/dotnet-interactive/udf_works.png?branch=pr-en-us-20778
Resource | docs/spark/how-to-guides/media/dotnet-interactive/udf-serialization-error.png | https://review.docs.microsoft.com/en-us/dotnet/spark/how-to-guides/media/dotnet-interactive/udf-serialization-error.png?branch=pr-en-us-20778
Resource | docs/spark/how-to-guides/media/invoke-external-udfs/JavaUdf.png | https://review.docs.microsoft.com/en-us/dotnet/spark/how-to-guides/media/invoke-external-udfs/javaudf.png?branch=pr-en-us-20778
Resource | docs/spark/how-to-guides/media/invoke-external-udfs/sparkdotnetapp.png | https://review.docs.microsoft.com/en-us/dotnet/spark/how-to-guides/media/invoke-external-udfs/sparkdotnetapp.png?branch=pr-en-us-20778

Commit:
Adding Spark dotnet how to guides

Commit:
fixing markdown errors

Commit:
markdown errors

Commit:
fixing markdown errors

Niharikadutta commented:
@luisquintanilla @rapoth @imback82 @suhsteve Kindly review this PR. Thank you!

Youssef1313 commented on docs/spark/how-to-guides/connect-to-azure-storage.md:
```diff
@@ -0,0 +1,74 @@
+    1. Set the `JAVA_HOME` property using the DOS path (since Hadoop doesn't like spaces in `JAVA_HOME`). This should look something like this: `C:\Progra~1\Java\jdk1.8.0_241` (Pointing to whatever version of Java you have installed on your local machine).
+    2. Append `%HADOOP_HOME%\share\hadoop\tools\lib\*` to `HADOOP_CLASSPATH`.
+    Your final `hadoop-env.cmd` file should look something like this:
+    ![Final hadoop-env.cmd file](./media/connect-external-sources/hadoop-env.png)
```
Images that aren't surrounded by blank lines used to cause problems in the localized docs.

```suggestion

    ![Final hadoop-env.cmd file](./media/connect-external-sources/hadoop-env.png)
```

Youssef1313 commented on docs/spark/how-to-guides/connect-to-azure-storage.md:
````diff
@@ -0,0 +1,74 @@
+
+This should display a list of all files/folders in the path provided by your URI.
+
+> NOTE: The format to derive the URI to your Storage account is as follows:
+
+```
+For ADLS: abfs[s]://<file_system>@<account_name>.dfs.core.windows.net/<path>/<file_name>
+For WASB: wasb[s]://<file_system>@<account_name>.blob.core.windows.net/<path>/<file_name>
+```
````
```suggestion
> [!NOTE]
> The format to derive the URI to your Storage account is as follows:
>
> ```
> For ADLS: abfs[s]://<file_system>@<account_name>.dfs.core.windows.net/<path>/<file_name>
> For WASB: wasb[s]://<file_system>@<account_name>.blob.core.windows.net/<path>/<file_name>
> ```
```

Youssef1313 commented on docs/spark/how-to-guides/connect-to-azure-storage.md:
````diff
@@ -0,0 +1,74 @@
+2. Set the output of the command run in Step 1 to the value of environment variable `SPARK_DIST_CLASSPATH`.
+3. Now you should be able to access your ADLS or WASB storage account through Spark .NET using the abfs URI as shown in the simple example below:
+
+  ```csharp
+  SparkSession spark = SparkSession
+      .Builder()
+      .AppName("Connect to Azure Storage locally")
+      .GetOrCreate();
+  DataFrame df = spark.Read().Json("wasbs://file_system@account_name.blob.core.windows.net/path/file.json");
+  //DataFrame df = spark.Read().Json("abfss://file_system@account_name.dfs.core.windows.net/path/file.json");
+  df.Show();
+  ```
````
The code block currently doesn't show **inside** the list item. This should fix that.

~~~suggestion
   ```csharp
   SparkSession spark = SparkSession
       .Builder()
       .AppName("Connect to Azure Storage locally")
       .GetOrCreate();
   DataFrame df = spark.Read().Json("wasbs://file_system@account_name.blob.core.windows.net/path/file.json");
   //DataFrame df = spark.Read().Json("abfss://file_system@account_name.dfs.core.windows.net/path/file.json");
   df.Show();
   ```
~~~

Youssef1313 commented on docs/spark/how-to-guides/connect-to-event-hub.md:
```diff
@@ -0,0 +1,83 @@
+
+## Steps to get your application connected to Event Hubs
+
+1. Create an Event Hubs Namespace and an event hub, refer to [this document](https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-create) for a step-by-step guide on how to do that. Make sure to select the Standard Pricing tier while creating the Event Hub Namespace.
```
Avoid localized links.

```suggestion
1. Create an Event Hubs Namespace and an event hub, refer to [this document](https://docs.microsoft.com/azure/event-hubs/event-hubs-create) for a step-by-step guide on how to do that. Make sure to select the Standard Pricing tier while creating the Event Hub Namespace.
```

Youssef1313 commented on docs/spark/how-to-guides/connect-to-event-hub.md:
```diff
@@ -0,0 +1,83 @@
+## Steps to get your application connected to Event Hubs
+
+1. Create an Event Hubs Namespace and an event hub, refer to [this document](https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-create) for a step-by-step guide on how to do that. Make sure to select the Standard Pricing tier while creating the Event Hub Namespace.
+2. Get the Event Hubs connection string and fully qualified domain name (FQDN) for later use. For instructions, see [Get an Event Hubs connection string](https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-get-connection-string).
```
```suggestion
2. Get the Event Hubs connection string and fully qualified domain name (FQDN) for later use. For instructions, see [Get an Event Hubs connection string](https://docs.microsoft.com/azure/event-hubs/event-hubs-get-connection-string).
```

Youssef1313 commented on docs/spark/how-to-guides/connect-to-event-hub.md:
```diff
@@ -0,0 +1,83 @@
+1. Create an Event Hubs Namespace and an event hub, refer to [this document](https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-create) for a step-by-step guide on how to do that. Make sure to select the Standard Pricing tier while creating the Event Hub Namespace.
+2. Get the Event Hubs connection string and fully qualified domain name (FQDN) for later use. For instructions, see [Get an Event Hubs connection string](https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-get-connection-string).
+3. Set the following configurations with details from your namespace in your code to start reading from Event Hubs for Kafka:
+    1. Update **BOOTSTRAP_SERVERS** to `hostname:9093` (9093 is the port used to communicate with Event Hubs, see [troubleshooting guide](https://docs.microsoft.com/en-us/azure/event-hubs/troubleshooting-guide))
```
```suggestion
    1. Update **BOOTSTRAP_SERVERS** to `hostname:9093` (9093 is the port used to communicate with Event Hubs, see [troubleshooting guide](https://docs.microsoft.com/azure/event-hubs/troubleshooting-guide))
```

Youssef1313 commented on docs/spark/how-to-guides/connect-to-event-hub.md:
````diff
@@ -0,0 +1,83 @@
+    1. Update **BOOTSTRAP_SERVERS** to `hostname:9093` (9093 is the port used to communicate with Event Hubs, see [troubleshooting guide](https://docs.microsoft.com/en-us/azure/event-hubs/troubleshooting-guide))
+    2. Update **EH_SASL** with the Connection string obtained from Step 2 like so:
+
+    ```csharp
+    string EH_SASL = "org.apache.kafka.common.security.plain.PlainLoginModule required username=\"$ConnectionString\" password=\"<CONNECTION_STRING>\";";
+    ```
````
~~~suggestion
       ```csharp
       string EH_SASL = "org.apache.kafka.common.security.plain.PlainLoginModule required username=\"$ConnectionString\" password=\"<CONNECTION_STRING>\";";
       ```
~~~

Youssef1313 commented on docs/spark/how-to-guides/connect-to-event-hub.md:
````diff
@@ -0,0 +1,83 @@
+        .Start();
+    ```
+
+6. In order to run your .NET for Apache Spark application, you should define the `spark-sql-kafka-0-10` module as part of the build definition in your Spark project, e.g. as a `libraryDependency` in `build.sbt` for sbt projects. For Spark environments such as `spark-submit` (or `spark-shell`) you should use the `--packages` command-line option like so:
````
The use of `e.g.` is discouraged by the current guidelines.

Youssef1313 commented on docs/spark/how-to-guides/connect-to-event-hub.md:
```diff
@@ -0,0 +1,83 @@
+
+## Next steps
+
+* [Connect your Apache Spark application with Azure Event Hubs](https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-kafka-spark-tutorial)
```
```suggestion
* [Connect your Apache Spark application with Azure Event Hubs](https://docs.microsoft.com/azure/event-hubs/event-hubs-kafka-spark-tutorial)
```

Youssef1313 commented on docs/spark/how-to-guides/connect-to-sql-server.md:
```diff
@@ -0,0 +1,67 @@
+3. Make sure the default port for SQL server 1433 is allowed through the firewall.
+4. Open SQL configure manager to enable TCP/IP through the network configuration as shown below:
+
+    ![SQL server TCP/IP enable](./media/connect-external-sources/SqlServerTCPIP.png)
```
```suggestion
    ![SQL server TCP/IP enable](./media/connect-external-sources/SqlServerTCPIP.png)

```

Youssef1313 commented on docs/spark/how-to-guides/dotnet-interactive-udf-issue.md:
```diff
@@ -0,0 +1,65 @@
+
+If we mark the [`SparSession`](https://github.com/dotnet/spark/blob/master/src/csharp/Microsoft.Spark/Sql/SparkSession.cs#L20) class as `[Serializable]`, we can get this to work, but this is not an ideal solution as we don't want to give the user the ability to serialize a SparkSession object, as that could lead to some very weird undesirable behavior. This is a known issue and being tracked [her ...
+
+These are a few important things to keep in mind while implementing UDFs in .NET for Apache Spark interactive experience (such as [Azure Synapse Notebooks](https://docs.microsoft.com/en-us/azure/synapse-analytics/spark/apache-spark-development-using-notebooks)).
```
```suggestion
These are a few important things to keep in mind while implementing UDFs in .NET for Apache Spark interactive experience (such as [Azure Synapse Notebooks](https://docs.microsoft.com/azure/synapse-analytics/spark/apache-spark-development-using-notebooks)).
```

Youssef1313 reviewed (commented):
Images in this PR doesn't follow the current folder structure guidelines.

See https://docs.microsoft.com/en-us/contribute/dotnet/dotnet-contribute#example-folder-structure

Commit:
Update docs/spark/how-to-guides/connect-to-azure-storage.md

Co-authored-by: Youssef Victor <31348972+Youssef1313@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-azure-storage.md

Co-authored-by: Youssef Victor <31348972+Youssef1313@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-azure-storage.md

Co-authored-by: Youssef Victor <31348972+Youssef1313@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-event-hub.md

Co-authored-by: Youssef Victor <31348972+Youssef1313@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-event-hub.md

Co-authored-by: Youssef Victor <31348972+Youssef1313@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-event-hub.md

Co-authored-by: Youssef Victor <31348972+Youssef1313@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-event-hub.md

Co-authored-by: Youssef Victor <31348972+Youssef1313@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-event-hub.md

Co-authored-by: Youssef Victor <31348972+Youssef1313@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-sql-server.md

Co-authored-by: Youssef Victor <31348972+Youssef1313@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/dotnet-interactive-udf-issue.md

Co-authored-by: Youssef Victor <31348972+Youssef1313@users.noreply.github.com>

Commit:
PR comments

Commit:
Merge branch 'nidutta/HowToGuides' of github.com:Niharikadutta/docs into nidutta/HowToGuides

Commit:
Forgot to add doc for mongodb

Commit:
markdown errors

Commit:
markdown errors

luisquintanilla commented on docs/spark/how-to-guides/connect-to-mongo-db.md:
```diff
@@ -0,0 +1,49 @@
+
+2. Use the `com.mongodb.spark.sql.DefaultSource` format is `spark.Read()` as shown below in a simple code snippet:
+
+    ![Sample MongoDB connection code](./media/connect-external-sources/mongo-code.png)
```
For this, would it be possible to include the code snippet instead of the screenshot?

Niharikadutta replied:
Done.

luisquintanilla commented on docs/spark/how-to-guides/connect-to-sql-server.md:
```diff
@@ -0,0 +1,68 @@
+
+# Connect to SQL server locally from .NET for Apache Spark
+
+In this article, you learn how to connect to an SQL server instance from your [.NET for Apache Spark](https://github.com/dotnet/spark) application.
```
```suggestion
In this article, you learn how to connect to a SQL server instance from your [.NET for Apache Spark](https://github.com/dotnet/spark) application.
```

Niharikadutta replied:
I think `an` is the correct usage in this case.

luisquintanilla replied:
@Niharikadutta Re: a/an I'd say it out loud and go with what sounds better. For me, a SQL server instance sounds more natural than an SQL server instance.

I also made a minor tweak to point to docs instead of the repo.

luisquintanilla commented on docs/spark/how-to-guides/connect-to-sql-server.md:
```diff
@@ -0,0 +1,68 @@
+
+In this article, you learn how to connect to an SQL server instance from your [.NET for Apache Spark](https://github.com/dotnet/spark) application.
+
+## Configure your SQL server to allow access from your .NET for Apache Spark application
```
```suggestion
## Configure SQL server to grant your application access
```

luisquintanilla commented on docs/spark/how-to-guides/connect-to-sql-server.md:
```diff
@@ -0,0 +1,68 @@
+
+    ![SQL server TCP/IP enable](./media/connect-external-sources/SqlServerTCPIP.png)
+
+    Also note the value of `Listen All` in above tab under `Protocol`.
```
```suggestion
    Also note the value of **Listen All** in above tab under **Protocol**.
```

luisquintanilla commented on docs/spark/how-to-guides/connect-to-sql-server.md:
```diff
@@ -0,0 +1,68 @@
+
+    Also note the value of `Listen All` in above tab under `Protocol`.
+
+5. Configure the TCP/IP port to 1433 for all required IP addresses if the `Listen All` is set to `No`. Otherwise, set the TCP Port in IPAll.
```
```suggestion
5. Configure the TCP/IP port to 1433 for all required IP addresses if the **Listen All** is set to **No**. Otherwise, set the TCP Port in IPAll.
```

luisquintanilla replied:
@Niharikadutta since these are UI elements and settings to look out for, it makes sense to highlight them

luisquintanilla commented on docs/spark/how-to-guides/connect-to-sql-server.md:
```diff
@@ -0,0 +1,68 @@
+
+    ![SQL server TCP/IP port](./media/connect-external-sources/SQLServerTCPIIPPort.png)
+
+## Configure your .NET for Apache Spark application to connect to the SQL server instance
```
```suggestion
## Connect to SQL Server from your application
```

luisquintanilla commented on docs/spark/how-to-guides/connect-to-sql-server.md:
```diff
@@ -0,0 +1,68 @@
+
+## Configure your .NET for Apache Spark application to connect to the SQL server instance
+
+1. You can use the Microsoft JDBC Driver for SQL server to provide database connectivity through your application (download from [this official website](https://docs.microsoft.com/en-us/sql/connect/jdbc/download-microsoft-jdbc-driver-for-sql-server?view=sql-server-ver15)).
```
```suggestion
1. Use the Microsoft JDBC Driver for SQL server to provide database connectivity through your application (download from [this official website](https://docs.microsoft.com/en-us/sql/connect/jdbc/download-microsoft-jdbc-driver-for-sql-server?view=sql-server-ver15)).
```

luisquintanilla commented on docs/spark/how-to-guides/dotnet-interactive-udf-issue.md:
```diff
@@ -0,0 +1,65 @@
+---
+title: Using UDFs in .NET for Apache Spark interactive shells.
```
```suggestion
title: Write UDFs in .NET for Apache Spark interactive environments.
```

luisquintanilla commented on docs/spark/how-to-guides/dotnet-interactive-udf-issue.md:
```diff
@@ -0,0 +1,65 @@
+
+# How to write UDFs in .NET for Apache Spark interactive shells
+
+In this article you will learn how to use UDFs in a .NET for Apache Spark interactive environment, a couple of important caveats to keep in mind, and why Broadcast Variables don't work with .NET for Apache Spark interactive experiences.
```
```suggestion
In this article you will learn how to use User Defined Functions (UDFs) in a .NET for Apache Spark interactive environment. 
```

luisquintanilla commented on docs/spark/how-to-guides/dotnet-interactive-udf-issue.md:
```diff
@@ -0,0 +1,65 @@
+ms.custom: mvc,how-to
+---
+
+# How to write UDFs in .NET for Apache Spark interactive shells
```
```suggestion
# How to write UDFs in .NET for Apache Spark interactive environments
```

luisquintanilla commented on docs/spark/how-to-guides/dotnet-interactive-udf-issue.md:
```diff
@@ -0,0 +1,65 @@
+
+## .NET for Apach Spark Interactive experience
+
+[.NET for Apache Spark](https://github.com/dotnet/spark) uses [.NET Interactive](https://devblogs.microsoft.com/dotnet/net-interactive-is-here-net-notebooks-preview-2/) to provide .NET support for interactive experience inside Spark. Kindly go through the [official docs](https://github.com/dotnet/interactive) to understand how to setup the environment to try .NET Interactive with Jupyter notebook ...
```
```suggestion
[.NET for Apache Spark](https://github.com/dotnet/spark) uses [.NET Interactive](https://devblogs.microsoft.com/dotnet/net-interactive-is-here-net-notebooks-preview-2/) to provide .NET support for interactive experience inside Spark. To understand how to setup the environment to try .NET Interactive with Jupyter notebooks, see the [.NET Interactive repository](https://github.com/dotnet/interactive).
```

luisquintanilla commented on docs/spark/how-to-guides/dotnet-interactive-udf-issue.md:
```diff
@@ -0,0 +1,65 @@
+[.NET for Apache Spark](https://github.com/dotnet/spark) uses [.NET Interactive](https://devblogs.microsoft.com/dotnet/net-interactive-is-here-net-notebooks-preview-2/) to provide .NET support for interactive experience inside Spark. Kindly go through the [official docs](https://github.com/dotnet/interactive) to understand how to setup the environment to try .NET Interactive with Jupyter notebook ...
+
+Also, it is recommended to go through [this article about UDF serialization in .NET for Apache Spark](udf-guide.md) to understand what UDFs are and how they are serialized in .NET for Apache Spark.
+We will be using Jupyter Notebooks to illustrate how to use UDFs in an interactive experience.
```
```suggestion
This guide uses Jupyter Notebooks to illustrate how to use UDFs in an interactive experience.
```

luisquintanilla commented on docs/spark/how-to-guides/dotnet-interactive-udf-issue.md:
```diff
@@ -0,0 +1,65 @@
+
+## UDF serialization in .NET Interactive
+
+In the interactive experience, a cell can be thought of as the code snippet being submitted as part of one iteration of the [REPL](https://en.wikipedia.org/wiki/Read%E2%80%93eval%E2%80%93print_loop). For example, let's take a look at the following notebook to understand what that means:
```
```suggestion
In the interactive experience, a cell can be thought of as the code snippet being submitted as part of one iteration of the [REPL](https://en.wikipedia.org/wiki/Read%E2%80%93eval%E2%80%93print_loop). For example, take a look at the following notebook to understand what that means:
```

luisquintanilla commented on docs/spark/how-to-guides/dotnet-interactive-udf-issue.md:
```diff
@@ -0,0 +1,65 @@
+
+![Jupyter Notebook cells](./media/dotnet-interactive/dotnet-interactive-cells.png)
+
+Each of those blocks marked by the red arrow is one cell, or one submission of code to Spark. .NET Interactive wraps each of these cells with a wrapper class of the cell submission number, to uniquely identify each cell being submitted. Now as explained in detail in [this guide](udf-guide.md), when a UDF that references a custom object is being serialized its target is also picked up for serializ ...
```
```suggestion
Each of those blocks marked by the red arrow is one cell, or one submission of code to Spark. .NET Interactive wraps each of these cells with a wrapper class of the cell submission number, to uniquely identify each cell being submitted. As explained in detail in [the UDF guide](udf-guide.md), when a UDF that references a custom object is serialized, its target is also picked up for serialization. In the case of .NET Interactive, it gets wrapped by the wrapper class of the cell where the custom object is defined.
```

luisquintanilla commented on docs/spark/how-to-guides/dotnet-interactive-udf-issue.md:
```diff
@@ -0,0 +1,65 @@
+
+![UDF serialization error](./media/dotnet-interactive/udf-serialization-error.png)
+
+As can be seen in the case of `udf2_fails`, we see the error message which says Type `Submission#7` is not marked as serializable, this is because how .NET Interactive works is it wraps every object defined in a cell with its `Submission#` class which is generated on the fly and hence is not marked as `Serializable`, hence the error.
```
See the second inline snippet `Submission#`. Is that right?

Niharikadutta replied:
Yes, it refers to the submission#<cell_number> compiler generated class that wraps the submission logic inside the cell.

luisquintanilla commented on docs/spark/how-to-guides/dotnet-interactive-udf-issue.md:
```diff
@@ -0,0 +1,65 @@
+![UDF serialization error](./media/dotnet-interactive/udf-serialization-error.png)
+
+As can be seen in the case of `udf2_fails`, we see the error message which says Type `Submission#7` is not marked as serializable, this is because how .NET Interactive works is it wraps every object defined in a cell with its `Submission#` class which is generated on the fly and hence is not marked as `Serializable`, hence the error.
+For this reason, it is required that a UDF referencing a custom object in it, is defined in the same cell as that object.
```
Highlight this point

```suggestion
For this reason, it is **required that a UDF referencing a custom object in it, is defined in the same cell as that object**.
```

luisquintanilla commented on docs/spark/how-to-guides/dotnet-interactive-udf-issue.md:
```diff
@@ -0,0 +1,65 @@
+
+## Why Broadcast Variables don't work with .NET Interactive
+
+The above explained behavior is the reason why Broadcast Variables don't work with .NET Interactive. It is a good idea to go through [this guide on Broadcast Variables](broadcast-guide.md) to get a deeper understanding of what Broadcast variables are and how to use them. The reason Broadcast Variables don't work with interactive scenarios is because of .NET interactive's design of appending each  ...
```
```suggestion
For the reasons explained above, broadcast variables don't work with .NET Interactive. It is a good idea to go through [this guide on broadcast variables](broadcast-guide.md) to get a deeper understanding of what broadcast variables are and how to use them. The reason broadcast variables don't work with interactive scenarios is because of .NET interactive's design of appending each object defined in a cell with it's cell submission class, which since is not marked serializable, fails with the same exception as shown previously.
```

luisquintanilla commented on docs/spark/how-to-guides/dotnet-interactive-udf-issue.md:
```diff
@@ -0,0 +1,65 @@
+
+![Broadcast Variables fail](./media/dotnet-interactive/broadcast-fails.png)
+
+As recommended in the previous sections, we define both the UDF and the object it is referencing (Broadcast variable in this case) in the same cell, but we still see the `SerializationException` error complaining about `Microsoft.Spark.Sql.Session` not being marked as serializable. This is because when the compiler tries to serialize the Broadcast variable object `bv`, it finds its name to be app ...
```
```suggestion
As recommended in the previous sections, we define both the UDF and the object it is referencing (broadcast variable in this case) in the same cell, but we still see the `SerializationException` error complaining about `Microsoft.Spark.Sql.Session` not being marked as serializable. This is because when the compiler tries to serialize the broadcast variable object `bv`, it finds its name to be appended with [`SparkSession`](https://github.com/dotnet/spark/blob/master/src/csharp/Microsoft.Spark/Sql/SparkSession.cs#L20) object `spark` which it requires to be marked as serializable. This can be demonstrated with more ease by taking a peek at the decompiled assembly of this cell submission:
```

luisquintanilla commented on docs/spark/how-to-guides/dotnet-interactive-udf-issue.md:
```diff
@@ -0,0 +1,65 @@
+
+![Decompiled Assembly code](./media/dotnet-interactive/decompiledAssembly.png)
+
+If we mark the [`SparSession`](https://github.com/dotnet/spark/blob/master/src/csharp/Microsoft.Spark/Sql/SparkSession.cs#L20) class as `[Serializable]`, we can get this to work, but this is not an ideal solution as we don't want to give the user the ability to serialize a SparkSession object, as that could lead to some very weird undesirable behavior. This is a known issue and being tracked [her ...
```
Not sure if we should link to the GitHub issue in docs. I'll check in on that.

luisquintanilla commented on docs/spark/how-to-guides/java-udf-from-dotnet.md:
```diff
@@ -0,0 +1,55 @@
+
+# Calling a Java UDF from your .NET for Apache Spark application
+
+In this article, you learn how to call a Java UDF (User-Defined Function) from your [.NET for Apache Spark](https://github.com/dotnet/spark) application. Let's go through the steps divided into the following sections:
```
```suggestion
In this article, you learn how to call a Java User-Defined Function (UDF) from your [.NET for Apache Spark](https://github.com/dotnet/spark) application. 
```

luisquintanilla reviewed (commented):
@Niharikadutta 

Looks great! Added a few suggestions. Let me know if you have questions.

Commit:
Update docs/spark/how-to-guides/connect-to-azure-storage.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-azure-storage.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-azure-storage.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-azure-storage.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/dotnet-interactive-udf-issue.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/dotnet-interactive-udf-issue.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/dotnet-interactive-udf-issue.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/java-udf-from-dotnet.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-event-hub.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-sql-server.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Fixing markdown errors

Niharikadutta commented on docs/spark/how-to-guides/connect-to-azure-storage.md:
```diff
@@ -0,0 +1,76 @@
+
+In this article, you learn how to connect to an ADLS (Azure Data Lake Storage) Gen 2 or WASB (Windows Azure Storage Blob) account through an instance of [.NET for Apache Spark](https://github.com/dotnet/spark) running locally on your Windows machine.
+
+## Steps to setup the environment
```
These steps are different from the generic setting up environment in the getting started guide since it requires installing Apache Spark without Hadoop and separate Hadoop installation.

Niharikadutta commented on docs/spark/how-to-guides/connect-to-event-hub.md:
```diff
@@ -0,0 +1,83 @@
+
+## Steps to get your application connected to Event Hubs
+
+1. Create an Event Hubs Namespace and an event hub, refer to [this document](https://docs.microsoft.com/azure/event-hubs/event-hubs-create) for a step-by-step guide on how to do that. Make sure to select the Standard Pricing tier while creating the Event Hub Namespace.
```
Makes sense. Moved to prerequisites section.

Niharikadutta commented on docs/spark/how-to-guides/connect-to-event-hub.md:
```diff
@@ -0,0 +1,83 @@
+
+[Azure Event Hubs](https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-about) is a big data streaming platform and event ingestion service. It is a fully managed Platform-as-a-Service (PaaS), that can easily integrate with [Apache Kafka](https://kafka.apache.org/) to give you the PaaS Kafka experience without having to manage, configure or run your own clusters.
+
+## Steps to get your application connected to Event Hubs
```
Fixed.

Niharikadutta commented on docs/spark/how-to-guides/connect-to-event-hub.md:
```diff
@@ -0,0 +1,83 @@
+## Steps to get your application connected to Event Hubs
+
+1. Create an Event Hubs Namespace and an event hub, refer to [this document](https://docs.microsoft.com/azure/event-hubs/event-hubs-create) for a step-by-step guide on how to do that. Make sure to select the Standard Pricing tier while creating the Event Hub Namespace.
+2. Get the Event Hubs connection string and fully qualified domain name (FQDN) for later use. For instructions, see [Get an Event Hubs connection string](https://docs.microsoft.com/azure/event-hubs/event-hubs-get-connection-string).
```
I feel like this belongs more in this section as it isn't a prerequisite per se but something that needs to be done to connect the Spark .NET application to Event Hubs. But please let me know if that doesn't make sense and I will move it to the prereqs section. Thanks!

Niharikadutta commented on docs/spark/how-to-guides/connect-to-event-hub.md:
```diff
@@ -0,0 +1,83 @@
+
+1. Create an Event Hubs Namespace and an event hub, refer to [this document](https://docs.microsoft.com/azure/event-hubs/event-hubs-create) for a step-by-step guide on how to do that. Make sure to select the Standard Pricing tier while creating the Event Hub Namespace.
+2. Get the Event Hubs connection string and fully qualified domain name (FQDN) for later use. For instructions, see [Get an Event Hubs connection string](https://docs.microsoft.com/azure/event-hubs/event-hubs-get-connection-string).
+3. Set the following configurations with details from your namespace in your code to start reading from Event Hubs for Kafka:
```
Added variables in a single code snippet.

Niharikadutta commented on docs/spark/how-to-guides/connect-to-event-hub.md:
````diff
@@ -0,0 +1,83 @@
+        string EH_SASL = "org.apache.kafka.common.security.plain.PlainLoginModule required username=\"$ConnectionString\" password=\"<CONNECTION_STRING>\";";
+        ```
+
+4. You can now start streaming with Event Hubs as you would with Kafka. Sample code as shown below:
````
Made the change.

Niharikadutta commented on docs/spark/how-to-guides/connect-to-event-hub.md:
````diff
@@ -0,0 +1,83 @@
+        .Start();
+    ```
+
+6. In order to run your .NET for Apache Spark application, you should define the `spark-sql-kafka-0-10` module as part of the build definition in your Spark project, using `libraryDependency` in `build.sbt` for sbt projects. For Spark environments such as `spark-submit` (or `spark-shell`) you should use the `--packages` command-line option like so:
````
This is pretty standard practice I think, depends on what project building tool they choose and there is plenty of documentation for this online.

Niharikadutta commented on docs/spark/how-to-guides/connect-to-mongo-db.md:
````diff
@@ -0,0 +1,49 @@
+    )
+    ```
+
+2. Make sure the IP address of the system your .NET for Apache Spark application is running on is whitelisted for the MongoDB server to connect to.
````
Added link to guide.

Commit:
PR comments

Commit:
markdown errors

Commit:
Update docs/spark/how-to-guides/connect-to-sql-server.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-sql-server.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-sql-server.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/dotnet-interactive-udf-issue.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/dotnet-interactive-udf-issue.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/dotnet-interactive-udf-issue.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/dotnet-interactive-udf-issue.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/dotnet-interactive-udf-issue.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/dotnet-interactive-udf-issue.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
markdown errors

Commit:
markdown errors

Niharikadutta commented:
@luisquintanilla Thank you for reviewing! I have addressed all your comments and made some changes based on the team's feedback. Please let me know if it looks ok, thank you!

luisquintanilla commented:
Thanks! I'll take a look :)

luisquintanilla commented on docs/spark/how-to-guides/connect-to-sql-server.md:
```diff
@@ -0,0 +1,68 @@
+
+In this article, you learn how to connect to an SQL server instance from your [.NET for Apache Spark](https://github.com/dotnet/spark) application.
+
+## Configure SQL server to grant your application access
```
```suggestion
## Configure SQL Server to grant your application access
```

luisquintanilla commented on docs/spark/how-to-guides/connect-to-sql-server.md:
```diff
@@ -0,0 +1,68 @@
+1. Add a Login user and password choosing SQL server authentication to your SQL server instance.
+2. Give that login user necessary permissions at the relevant database level as shown below:
+
+    ![SQL server permissions](./media/connect-external-sources/SqlServerAuth.png)
```
```suggestion
    ![SQL Server permissions](./media/connect-external-sources/SqlServerAuth.png)
```

luisquintanilla commented on docs/spark/how-to-guides/connect-to-sql-server.md:
```diff
@@ -0,0 +1,68 @@
+
+## Configure SQL server to grant your application access
+
+1. Add a Login user and password choosing SQL server authentication to your SQL server instance.
```
```suggestion
1. Add a login user and password choosing SQL Server authentication to your SQL Server instance.
```

luisquintanilla commented on docs/spark/how-to-guides/connect-to-sql-server.md:
```diff
@@ -0,0 +1,68 @@
+
+    ![SQL server permissions](./media/connect-external-sources/SqlServerAuth.png)
+
+3. Make sure the default port for SQL server 1433 is allowed through the firewall.
```
```suggestion
3. Make sure the default port for SQL Server `1433` is allowed through the firewall.
```

luisquintanilla commented on docs/spark/how-to-guides/connect-to-sql-server.md:
```diff
@@ -0,0 +1,68 @@
+3. Make sure the default port for SQL server 1433 is allowed through the firewall.
+4. Open SQL configure manager to enable TCP/IP through the network configuration as shown below:
+
+    ![SQL server TCP/IP enable](./media/connect-external-sources/SqlServerTCPIP.png)
```
```suggestion
    ![SQL Server TCP/IP enable](./media/connect-external-sources/SqlServerTCPIP.png)
```

luisquintanilla commented on docs/spark/how-to-guides/connect-to-sql-server.md:
```diff
@@ -0,0 +1,68 @@
+---
+title: Connect to SQL server locally from .NET for Apache Spark
+description: Learn how to connect to an SQL server instance from your .NET for Apache Spark application.
+ms.date: 09/24/2020
```
```suggestion
ms.date: 10/09/2020
```

luisquintanilla commented on docs/spark/how-to-guides/connect-to-sql-server.md:
```diff
@@ -0,0 +1,68 @@
+---
+title: Connect to SQL server locally from .NET for Apache Spark
+description: Learn how to connect to an SQL server instance from your .NET for Apache Spark application.
```
```suggestion
description: Learn how to connect to a SQL Server instance from your .NET for Apache Spark application.
```

luisquintanilla commented on docs/spark/how-to-guides/connect-to-sql-server.md:
```diff
@@ -0,0 +1,68 @@
+
+## Connect to SQL Server from your application
+
+1. Use the Microsoft JDBC Driver for SQL server to provide database connectivity through your application (download from [this official website](https://docs.microsoft.com/en-us/sql/connect/jdbc/download-microsoft-jdbc-driver-for-sql-server?view=sql-server-ver15)).
```
```suggestion
1. Use the Microsoft JDBC Driver for SQL Server to provide database connectivity through your application (download from [this official website](https://docs.microsoft.com/sql/connect/jdbc/download-microsoft-jdbc-driver-for-sql-server?view=sql-server-ver15)).
```

luisquintanilla commented on docs/spark/how-to-guides/connect-to-sql-server.md:
```diff
@@ -0,0 +1,68 @@
+
+5. Configure the TCP/IP port to 1433 for all required IP addresses if the `Listen All` is set to `No`. Otherwise, set the TCP Port in IPAll.
+
+    ![SQL server TCP/IP port](./media/connect-external-sources/SQLServerTCPIIPPort.png)
```
```suggestion
    ![SQL Server TCP/IP port](./media/connect-external-sources/SQLServerTCPIIPPort.png)
```

luisquintanilla commented on docs/spark/how-to-guides/dotnet-interactive-udf-issue.md:
```diff
@@ -0,0 +1,68 @@
+---
+title: Write UDFs in .NET for Apache Spark interactive environments.
+description: Learn how to write and call UDFs in .NET for Apache Spark interactive shells.
+ms.date: 09/24/2020
```
```suggestion
ms.date: 10/09/2020
```

luisquintanilla commented on docs/spark/how-to-guides/dotnet-interactive-udf-issue.md:
```diff
@@ -0,0 +1,68 @@
+Also, it is recommended to go through [this article about UDF serialization in .NET for Apache Spark](udf-guide.md) to understand what UDFs are and how they are serialized in .NET for Apache Spark.
+This guide uses Jupyter Notebooks to illustrate how to use UDFs in an interactive experience.
+
+## Defining a User-defined Function (UDF) in .NET Interactive
```
```suggestion
## Define a UDF in .NET Interactive
```

luisquintanilla commented on docs/spark/how-to-guides/dotnet-interactive-udf-issue.md:
```diff
@@ -0,0 +1,68 @@
+
+![Working UDF](./media/dotnet-interactive/working-udf.png)
+
+## Calling a UDF on a DataFrame
```
```suggestion
## Call a UDF on a DataFrame
```

luisquintanilla commented on docs/spark/how-to-guides/dotnet-interactive-udf-issue.md:
```diff
@@ -0,0 +1,68 @@
+ms.custom: mvc,how-to
+---
+
+# How to write UDFs in .NET for Apache Spark interactive environments
```
```suggestion
# Write and call UDFs in .NET for Apache Spark interactive environments
```

luisquintanilla commented on docs/spark/how-to-guides/dotnet-interactive-udf-issue.md:
```diff
@@ -0,0 +1,68 @@
+---
+title: Write UDFs in .NET for Apache Spark interactive environments.
```
```suggestion
title: Write and call UDFs in .NET for Apache Spark interactive environments.
```

luisquintanilla commented on docs/spark/how-to-guides/java-udf-from-dotnet.md:
```diff
@@ -0,0 +1,55 @@
+---
+title: Invoking Java UDFs from .NET for Apache Spark application
```
```suggestion
title: Invoke Java UDFs from .NET for Apache Spark application
```

luisquintanilla commented on docs/spark/how-to-guides/java-udf-from-dotnet.md:
```diff
@@ -0,0 +1,55 @@
+---
+title: Invoking Java UDFs from .NET for Apache Spark application
+description: Learn how to call a Java UDF from a .NET for Apache Spark application.
+ms.date: 09/25/2020
```
```suggestion
ms.date: 10/09/2020
```

luisquintanilla commented on docs/spark/how-to-guides/java-udf-from-dotnet.md:
```diff
@@ -0,0 +1,55 @@
+ms.custom: mvc,how-to
+---
+
+# Calling a Java UDF from your .NET for Apache Spark application
```
```suggestion
# Call a Java UDF from your .NET for Apache Spark application
```

luisquintanilla commented on docs/spark/how-to-guides/java-udf-from-dotnet.md:
```diff
@@ -0,0 +1,55 @@
+
+## Calling a .NET UDF from Scala or Python in an Apache Spark applications
+
+You can also register and invoke a C# UDF from an Apache Spark application written in Scala or Python using this [open source tool](https://github.com/imback82/sparkdotnetudf).
```
```suggestion
You can also register and invoke a C# UDF from an Apache Spark application written in Scala or Python using [the sparkdotnetudf open source tool](https://github.com/imback82/sparkdotnetudf).
```

luisquintanilla commented on docs/spark/how-to-guides/java-udf-from-dotnet.md:
````diff
@@ -0,0 +1,55 @@
+    +-------+
+```
+
+## Calling a .NET UDF from Scala or Python in an Apache Spark applications
````
```suggestion
## Call .NET UDF from Scala or Python in Apache Spark
```

luisquintanilla commented on docs/spark/how-to-guides/java-udf-from-dotnet.md:
````diff
@@ -0,0 +1,55 @@
+    spark-submit --master local --jars UdfApp-0.0.1.jar --class org.apache.spark.deploy.dotnet.DotnetRunner microsoft-spark-3.0.x-0.12.1.jar InterRuntimeUDFs.exe
+    ```
+
+The resultant `dfUdf` DataFrame had the number 5 added to each row of the input column as defined by `JavaUdf`:
````
Formatting look s a bit off. Indent this sentence and code snippet below to fix it

Niharikadutta replied:
Fixed.

luisquintanilla commented on docs/spark/how-to-guides/java-udf-from-dotnet.md:
```diff
@@ -0,0 +1,55 @@
+1. How to define your Java UDFs and compile them into a jar - this step is not needed if you already have a UDF defined in a jar file. In which case, all you need is the full name of the UDF function including the package.
+2. Register and call your Java UDF in your .NET for Apache Spark application.
+
+## Define your Java UDF(s) and compile it into a jar
```
```suggestion
## Define and compile your Java UDFs
```

luisquintanilla commented on docs/spark/how-to-guides/java-udf-from-dotnet.md:
```diff
@@ -0,0 +1,55 @@
+
+3. Compile and package your project to create and executable jar say `UdfApp-0.0.1.jar`.
+
+## Register and call your Java UDF in your .NET for Apache Spark application
```
```suggestion
## Register and call Java UDFs in .NET for Apache Spark
```

luisquintanilla reviewed (approved):
@Niharikadutta please take a look at the comments. I've made them suggestions so you should be able to just commit if you agree. Approving.

Commit:
Update docs/spark/how-to-guides/connect-to-event-hub.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-azure-storage.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-event-hub.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-azure-storage.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-event-hub.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-event-hub.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-event-hub.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-mongo-db.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-mongo-db.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-mongo-db.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-mongo-db.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-sql-server.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-sql-server.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-sql-server.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-sql-server.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-sql-server.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-sql-server.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-sql-server.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-sql-server.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-sql-server.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-sql-server.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/connect-to-sql-server.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/dotnet-interactive-udf-issue.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/dotnet-interactive-udf-issue.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/dotnet-interactive-udf-issue.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/dotnet-interactive-udf-issue.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/dotnet-interactive-udf-issue.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/java-udf-from-dotnet.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/java-udf-from-dotnet.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/java-udf-from-dotnet.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/java-udf-from-dotnet.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/java-udf-from-dotnet.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/java-udf-from-dotnet.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Commit:
Update docs/spark/how-to-guides/java-udf-from-dotnet.md

Co-authored-by: Luis Quintanilla <46974588+luisquintanilla@users.noreply.github.com>

Niharikadutta commented on docs/spark/how-to-guides/connect-to-event-hub.md:
```diff
@@ -0,0 +1,92 @@
+
+## Next steps
+
+* [Connect your Apache Spark application with Azure Event Hubs](https://docs.microsoft.com/azure/event-hubs/event-hubs-kafka-spark-tutorial)
```
Makes sense, there is no solid reason so removing `Next steps` section here

Commit:
PR comments

Niharikadutta commented on docs/spark/how-to-guides/connect-to-mongo-db.md:
````diff
@@ -0,0 +1,93 @@
+
+2. Use the `com.mongodb.spark.sql.DefaultSource` format is `spark.Read()` as shown below in a simple code snippet:
+
+```csharp
````
Fixed.

Niharikadutta commented on docs/spark/how-to-guides/connect-to-azure-storage.md:
````diff
@@ -0,0 +1,88 @@
+   df.Show();
+   ```
+
+The result as displayed is the DataFrame (`df`) as shown below:
````
Fixed.

Niharikadutta commented on docs/spark/how-to-guides/connect-to-azure-storage.md:
````diff
@@ -0,0 +1,88 @@
+    If you are trying to connect to a WASB account, replace `dfs` with `blob` in the property names. For example, `fs.azure.account.auth.type.YOUR_ACCOUNT_NAME.blob.core.windows.net`.
+3. You can test the connectivity to your Storage Account from Hadoop by running the following command from your `%HADOOP_HOME%` directory:
+
+  ```bash
````
Fixed.

Niharikadutta commented on docs/spark/how-to-guides/connect-to-azure-storage.md:
```diff
@@ -0,0 +1,88 @@
+
+    ![Final hadoop-env.cmd file](./media/connect-external-sources/hadoop-env.png)
+
+## Configure Hadoop to connect to your storage account
```
Changed.

Commit:
markdown error

Commit:
formatting

Commit:
formatting

Commit:
formatting

Niharikadutta commented:
@luisquintanilla I have addressed all remaining comments, thanks again for reviewing!

imback82 commented:
Thanks @Niharikadutta and @luisquintanilla!
